runtime/drivers/pinot/driver.go

```diff
@@ -0,0 +1,282 @@
+package pinot
+
+import (
+	"context"
+	"database/sql"
+	sqlDriver "database/sql/driver"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"math/big"
+	"net/url"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/startreedata/pinot-client-go/pinot"
+)
+
+type pinotDriver struct{}
+
+func (d *pinotDriver) Open(dsn string) (sqlDriver.Conn, error) {
+	// validate dsn - it should be a valid URL, may contain basic auth credentials
+	u, err := url.Parse(dsn)
+	if err != nil {
+		return nil, fmt.Errorf("invalid DSN: %w", err)
+	}
+
+	var authHeader map[string]string
+	if u.User != nil {
+		uname := u.User.Username()
+		pwd, passwordSet := u.User.Password()
+		if uname == "" || !passwordSet {
+			return nil, fmt.Errorf("DSN should contain valid basic auth credentials")
+		}
+		// clear user info from URL so that u.String() doesn't include it
+		u.User = nil
+		authString := fmt.Sprintf("%s:%s", uname, pwd)
+		authHeader = map[string]string{
+			"Authorization": fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(authString))),
+		}
+	}
+
+	pinotConn, err := pinot.NewWithConfig(&pinot.ClientConfig{
+		ExtraHTTPHeader: authHeader,
+		ControllerConfig: &pinot.ControllerConfig{
+			ControllerAddress: u.String(),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	// TODO check if it needs to be configurable
+	pinotConn.UseMultistageEngine(true)
+	return &conn{pinotConn: pinotConn}, nil
+}
+
+func init() {
+	sql.Register("pinot", &pinotDriver{})
+}
+
+type conn struct {
+	pinotConn *pinot.Connection
+}
+
+func (c *conn) Prepare(query string) (sqlDriver.Stmt, error) {
+	return nil, fmt.Errorf("unsupported")
+}
+
+func (c *conn) Close() error {
+	return nil
+}
+
+func (c *conn) Begin() (sqlDriver.Tx, error) {
+	return nil, fmt.Errorf("unsupported")
+}
+
+func (c *conn) QueryContext(ctx context.Context, query string, args []sqlDriver.NamedValue) (sqlDriver.Rows, error) {
+	if len(args) > 0 {
+		q, err := completeQuery(query, args)
+		if err != nil {
+			return nil, err
+		}
+		query = q
+	}
+	// TODO: cancel the query if ctx is done
+	resp, err := c.pinotConn.ExecuteSQL("", query)
```

Comment:
https://docs.pinot.apache.org/basics/getting-started/frequent-questions/query-faq#how-do-i-cancel-a-query Will check if it can be added quickly, by default query cancellation is off

Reply:
Can you add to our Pinot docs that people need to enable these Pinot configs?